pkg/rsstorage/servers/file/server.go

```diff
@@ -127,28 +139,64 @@ func (s *StorageServer) CalculateUsage() (types.Usage, error) {
 
 // diskUsage will walk the specified path in a filesystem and
 // aggregate the size of the contained files.
-func diskUsage(duPath string, cacheTimeout time.Duration) (size datasize.ByteSize, err error) {
-	timeout := time.Now().Add(cacheTimeout)
-	sizep := &size
+func diskUsage(duPath string, cacheTimeout, walkTimeout time.Duration) (size datasize.ByteSize, err error) {
+	stop := make(chan bool)
+	defer close(stop)
+
+	sizeChan := make(chan datasize.ByteSize)
+	// errChan should have a buffer of two items to prevent deadlock between `<-stop` and `errChan<-err`
+	errChan := make(chan error, 2)
+
+	go func(stop <-chan bool, sizeChan chan<- datasize.ByteSize, errChan chan<- error) {
+		err = filepath.Walk(duPath, func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+
+			select {
+			case <-stop:
+				return nil
+			default:
+			}
+
+			if !info.IsDir() {
+				sizeChan <- datasize.ByteSize(info.Size())
+			}
+
+			return nil
+		})
 
-	err = filepath.Walk(duPath, func(path string, info os.FileInfo, err error) error {
 		if err != nil {
-			return err
+			errChan <- err
 		}
 
-		if time.Now().After(timeout) {
-			return errors.New("timeout in diskUsage")
-		}
+		close(sizeChan)
+		close(errChan)
+	}(stop, sizeChan, errChan)
 
-		if !info.IsDir() {
-			*sizep += datasize.ByteSize(info.Size())
-		}
+	cacheTimeoutTimer := time.NewTimer(cacheTimeout)
+	defer cacheTimeoutTimer.Stop()
 
-		return nil
-	})
-	size = *sizep
+	start := time.Now()
 
-	return
+	for {
+		if walkTimeout > 0 && time.Now().Sub(start) > walkTimeout {
+			return 0, walktimeoutErr
```

Comment:
I may be missing something here, but could we simply establish a `walkTimeoutTimer` above and add it to the `select` below to eliminate this conditional?

Reply:
Oh there's a `.Reset()` function isn't there? I was trying to avoid the `time.After(...)` memory leak, but totally forgot about this functionality.